backend/infrahub/core/utils.py

```diff
@@ -191,6 +191,68 @@ def convert_ip_to_binary_str(
     return ip_bin.zfill(obj.max_prefixlen)
 
 
+def collapse_ipv6(s: str) -> str:
+    """Collapse an ipv6 address, ipv6 network, or a partial ipv6 address. Raises an error if input does not resemble an IPv6 address."""
+
+    try:
+        return str(ipaddress.IPv6Address(s))
+    except ipaddress.AddressValueError:
+        pass
+
+    try:
+        return ipaddress.IPv6Network(s).with_prefixlen
+    except ipaddress.AddressValueError:
+        pass
+
+    # Input string might be an incomplete address in IPv6 format,
+    # in which case we would like the collapsed form equivalent of this incomplete address for matching purposes.
+    # To get it, we first try to pad the incomplete address with zeros, then we retrieve the collapsed form
+    # of the full address, and we remove extra "::" or ":0" at the end of it.
+
+    error_message = "Input string does not match IPv6 format"
+
+    if "::" in s:
+        raise ValueError(error_message)
```

Comment:
This is a wrong assumption. The compressed form of a 0 address in IPv6 is `::`.
```
>>> a = ipaddress.IPv6Address('::')
>>> a.compressed
'::'
>>> a.exploded
'0000:0000:0000:0000:0000:0000:0000:0000'
```

Also it's not illegal to have `::` in the middle of a compressed notation such as:
```
>>> a = ipaddress.IPv6Address('2001:db8::1')
>>> a.compressed
'2001:db8::1'
>>> a.exploded
'2001:0db8:0000:0000:0000:0000:0000:0001'
```

Reply:
Yes, the point of this `if` block is that if this condition is met, then input string cannot be an IPv6 in extended format, so we raise an error because we do not want to process it anymore (if input is the start of an IPv6 in collapsed format, there is nothing to do). That said this function deserves more documentation regarding this behavior and current `error_message` is misleading, I will clarify this.